designs/workflow-refactor.md

```diff
@@ -0,0 +1,296 @@
+
+#### Workload cluster workflows
+
+##### Create
+
+1. **Pre Create** executes pre workflow hooks.
+2. **Create Workload Cluster** creates the cluster by asking the provider to convert EKS-A APIs to APIs understood by cluster management components resulting in a functional cluster.
+3. **Install Networking** install the network configuration to support cluster communication.
+4. **Install EKSA Components** installs EKS-A types used to identify properties of the management cluster in subsequent operations.
+5. **Post Create** executes post workflow hooks.
+
+##### Upgrade
+
+1. **Pre Upgrade** executes pre workflow hooks.
+2. **Upgrade Networking** upgrades the networking implementation in the management cluster.
+3. **Upgrade EKSA Components** upgrades EKS-A specific components installed during the create *Install EKSA Components* task.
+4. **Upgrade Cluster** upgrades the management cluster’s by invoking the provider upgrade cluster behavior.
+5. **Write Cluster Config** writes an updated cluster configuration file to disk. This is necessary for upgrade operations as we may upgrade the EKS-A cluster configuration objects and they need to be made available to the user.
+6. **Post Upgrade** executes post workflow hooks.
+
+##### Delete
+
+Deletion of a workload cluster entails deletion of resources on the management cluster to invoke CAPI de-provisioning of nodes.
+
+#### Impact
+
+* Having separate workflows creates lower cognitive load as both workflow maintainers and provider developers have to hold less context mentally. 
+* The removal of tasks unrelated to producing a functional cluster further reduces cognitive load. 
+* It allows each type of workflow set to define the interface they expect from providers independently of other workflows. For example management workflows may require bootstrap related cluster management component installation methods that are only consumed by management workflows making it clear to a provider the context under which the method is called so they can make any necessary adjustments.
+* A document and clear set of guidelines for workflow maintainers on what granularity of task and types of task should feature explicitly in a workflow.
+
+### Separate workflow construction logic from task behavior
+
+Tasks are currently written to know whether or not they should run at the time they are invoked. They then return the next task to be run creating a directed acyclic graph (DAG) of tasks. 
+
+Instead of having tasks dictate the ordering, we will consolidate what tasks run in construction logic as part of the `management.NewCreateWorkflow()` set of functions. This will yield a single location for understanding the conditions for running a task, separate the concern of constructing a set of tasks from the behavior of a task, reduce the need to pass information in the context between tasks, and enable explicit declaration of task dependencies that can be injected into the `New*()` funcs.
```

Comment:
@jiayiwang7 talked about rollback functionality, and it can probably be defined through a reverse rollback workflow where it is a different set of functions mapped from a specific task point. Is this something we want to mention in this proposal as providing the option to achieve that?

Reply:
I think that's more a conversation piece given we don't have any rollback today.

Reply:
Decided to call out the forward thinking in the proposal.